server/src/main/java/io/crate/types/TimeType.java

```diff
@@ -0,0 +1,277 @@
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.types;
+
+import io.crate.Streamer;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.format.DateTimeParseException;
+import java.time.format.ResolverStyle;
+import java.time.temporal.ChronoField;
+import java.util.Locale;
+import java.util.function.Supplier;
+
+/**
+ * Represents time as microseconds from midnight, ignoring the time
+ * zone and storing the value as UTC <b>long</b>.
+ *
+ * <p>
+ *
+ * There are 1000_000 microseconds in one second:
+ *
+ * <pre>
+ *     (24 * 3600 + 59 * 60 + 59) * 1000_000L > Integer.MAX_VALUE
+ * </pre>
+ *
+ * Thus the range for time values is 0 .. 86400000000 (max number
+ * of micros in a day), where both extremes are equivalent to
+ * '00:00:00:000000' and '24:00:00.000000' respectively.
+ *
+ * <p>
+ *
+ * Accepts four kinds of literal:
+ * <ol>
+ *    <li>text:
+ *      <ul>
+ *        <li>'hhmmss': e.g. '232121', equivalent to '23:12:21'</li>
+ *        <li>'hhmm': e.g. '2312', equivalent to '23:12:00'</li>
+ *        <li>'hh': e.g. '23', equivalent to '23:00:00'</li>
+ *      </ul>
+ *    </li>
+ *
+ *    <li>numeric:
+ *      <ul>
+ *        <li>integer and long values are first interpreted as
+ *        'text'. Failing this they are kept as is, representing
+ *        microseconds from midnight, ignoring the time zone
+ *        and storing the value as UTC.
+ *    </li>
+ *
+ *    <li>text high precision:
+ *      <p>
+ *      Expects up to six digits after the floating point (number of
+ *      micro seconds), and it will right pad with zeroes if this is
+ *      not the case. For instance the examples below are all padded
+ *      to 999000 micro seconds.
+ *      <ul>
+ *        <li>'hhmmss.ffffff': e.g. '231221.999', equivalent to '23:12:21.999'</li>
+ *        <li>'hhmm.ffffff': e.g. '2312.999', equivalent to '23:12:00.999'</li>
+ *        <li>'hh.ffffff': e.g. '23.999', equivalent to '23:00:00.999'</li>
+ *      </ul>
+ *    </li>
+ *
+ *    <li>numeric high precision:
+ *      <ul>
+ *        <li>double and float values are interpreted as
```

Comment:
PostgreSQL doesn't support numeric conversion to time:

```
select 231221.999::time;
ERROR:  42846: cannot cast type numeric to time without time zone
LINE 1: select 231221.999::time
```

I think we should also restrict this and not support it.

Compared to using a string literal, which is supported:

```
# select '231221.999'::time;
     time
--------------
 23:12:21.999
(1 row)
```